agent/config/file_watcher.go

```diff
@@ -0,0 +1,222 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-hclog"
+)
+
+const timeoutDuration = 200 * time.Millisecond
+
+type FileWatcher struct {
+	watcher          *fsnotify.Watcher
+	configFiles      map[string]*watchedFile
+	EventsCh         chan *WatcherEvent
```

Comment:
Should document this public field. It can be useful to keep the exported and un-exported struct members separated out.

Alternatively you could leave this one unexported, and provide an accessor method that forces it to be a receive-only channel:

```
...
    eventsCh chan *WatcherEvent
...
func (w *FileWatcher) EventsCh() <-chan *WatcherEvent { return w.eventsCh }
```

Reply:
Should also (like Start/Stop) describe how to correctly use this channel to avoid blocking this detection loop, something like "be prompt in removing items from the channel" or something to that effect.